### PR TITLE
Lingo: Add panels mode door shuffle

### PR DESCRIPTION
**Pull request by hatkirby**
## What is this fixing or adding?
This adds a new type of door shuffle called "panels mode". It is similar to the panels mode door shuffle in The Witness. Shuffle Doors can be set to none, panels mode, or doors mode. A new option called "group doors" has also been added. Simple doors and complex doors from before are now achieved by selecting doors mode, and then enabling or disabling "group doors".

Panels mode retains the vanilla behaviour of the doors in the game, but many panels are now locked behind key items. This adds a lot of progression blocking, similar to (though not as much as) doors mode door shuffle, but retains a more vanilla experience, because doors must still be opened from the correct side. This can interact with painting shuffle and sunwarp shuffle to create new, bizarre paths through the game.

## How was this tested?
pytest, test generations, a couple of test playthroughs.

Prerelease versions of the client and tracker are available for testing too:

Apworld: [lingo.zip](https://github.com/user-attachments/files/16362981/lingo.zip) (Rename to lingo.apworld)
Client: [lingo-archipelago-panels-beta6.zip](https://github.com/user-attachments/files/16362989/lingo-archipelago-panels-beta6.zip)
Tracker: [lingo-ap-tracker-panels-beta6-win64.zip](https://github.com/user-attachments/files/16362984/lingo-ap-tracker-panels-beta6-win64.zip)

### CODE REVIEW

**Comment by hatkirby** [comment]
Merged in the Pilgrim Update and updated the client and tracker accordingly.

**Comment by hatkirby** [comment]
Updated client and tracker with new changes.

**Review by NewSoupVi — APPROVED** [review]
Looks good but it's hard to review, we should playtest this some time

File: worlds/lingo/player_logic.py [hunk 1]
```diff
@@ -294,15 +328,14 @@ def __init__(self, world: "LingoWorld"):
                 # HOT CRUST and THIS.
                 good_item_options.append("Pilgrim Room - Sun Painting")
 
-            if not color_shuffle:
```
[the remark this answers is not in the document]
Comment (hatkirby) [reply worlds/lingo/player_logic.py:297]: This was a merge error; resolved.

File: worlds/lingo/player_logic.py [hunk 2]
```diff
@@ -420,12 +451,30 @@ def calculate_panel_requirements(self, room: str, panel: str, world: "LingoWorld
             access_reqs = AccessRequirements()
             panel_object = PANELS_BY_ROOM[room][panel]
 
+            if world.options.shuffle_doors == ShuffleDoors.option_panels and panel_object.panel_door is not None:
+                pd_ref = panel_object.panel_door
```
[the remark this answers is not in the document]
Comment (hatkirby) [reply worlds/lingo/player_logic.py]: Rewrote that section a little bit. Is this better?

**Comment by NewSoupVi** [comment]
Currently has merge conflicts, also I'd definitely like another non-me review on this

File: worlds/lingo/player_logic.py [hunk 3]
```diff
@@ -153,17 +162,27 @@ def __init__(self, world: "LingoWorld"):
...
-        if location_checks == LocationChecks.option_reduced and door_shuffle != ShuffleDoors.option_none:
-            raise OptionError("You cannot have reduced location checks when door shuffle is on, because there would not"
-                              " be enough locations for all of the door items.")
+        if location_checks == LocationChecks.option_reduced:
+            if door_shuffle == ShuffleDoors.option_doors:
+                raise OptionError("You cannot have reduced location checks when door shuffle is on, because there "
+                                  "would not be enough locations for all of the door items.")
+            if door_shuffle == ShuffleDoors.option_panels:
+                if not world.options.group_doors:
+                    raise OptionError("You cannot have reduced location checks when complex panels mode door shuffle "
```
Comment (Exempt-Medic) [inline worlds/lingo/player_logic.py:171]: This isn't called complex anymore

File: worlds/lingo/player_logic.py [hunk 4]
```diff
@@ -153,17 +162,29 @@ def __init__(self, world: "LingoWorld"):
         victory_condition = world.options.victory_condition
         early_color_hallways = world.options.early_color_hallways
 
-        if location_checks == LocationChecks.option_reduced and door_shuffle != ShuffleDoors.option_none:
-            raise OptionError("You cannot have reduced location checks when door shuffle is on, because there would not"
-                              " be enough locations for all of the door items.")
+        if location_checks == LocationChecks.option_reduced:
+            if door_shuffle == ShuffleDoors.option_doors:
+                raise OptionError("You cannot have reduced location checks when door shuffle is on, because there "
```
Comment (Exempt-Medic) [inline worlds/lingo/player_logic.py:167]: These warning should include the player name

**Review by Exempt-Medic — APPROVED** [review]
Did not read through `.rb` changes or `pickle_static_data.py`. The rest of the changes LGTM. I crawled through `LL1.yaml` and the changes there all made sense. The tests pass and what they are testing matches my understanding of the game. Looked through the `options.py` changes by checking them on WebHost as well, and searched through for instances of `ShuffleDoors` to check that they were updated properly with the option change. Merged into `main` and did 5000 generations. Played through 20 seeds and walked through 12 spoiler logs / playthroughs. Confirmed that the new option warnings function properly.

One comment on the option error display, only other thing is that I wonder if [this section](https://github.com/ArchipelagoMW/Archipelago/blob/6dd489dac2841f97d076dc34d80a59c6db5394bb/worlds/lingo/player_logic.py#L216-L228) and [this section](https://github.com/ArchipelagoMW/Archipelago/blob/6dd489dac2841f97d076dc34d80a59c6db5394bb/worlds/lingo/player_logic.py#L493-L505) could be turned into a function instead of semi-duplicated code

File: worlds/lingo/player_logic.py (hunk 4 above)
Comment (hatkirby) [reply worlds/lingo/player_logic.py]: Done!

**Pull request merged**